library/rsa.c

```diff
@@ -539,26 +542,32 @@ int mbedtls_rsa_gen_key( mbedtls_rsa_context *ctx,
         MBEDTLS_MPI_CHK( mbedtls_mpi_sub_int( &ctx->P, &ctx->P, 1 ) );
         MBEDTLS_MPI_CHK( mbedtls_mpi_sub_int( &ctx->Q, &ctx->Q, 1 ) );
         MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &H, &ctx->P, &ctx->Q ) );
+
+        /* check GCD( E, (P-1)*(Q-1) ) == 1 */
         MBEDTLS_MPI_CHK( mbedtls_mpi_gcd( &G, &ctx->E, &H  ) );
+        if ( mbedtls_mpi_cmp_int( &G, 1 ) != 0 ) {
+            continue;
+        }
+
+        /* compute smallest possible D = E^-1 mod LCM(P-1, Q-1) */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_gcd( &G, &ctx->P, &ctx->Q ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_div_mpi( &L, NULL, &H, &G ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_inv_mod( &ctx->D, &ctx->E, &L ) );
     }
-    while( mbedtls_mpi_cmp_int( &G, 1 ) != 0 );
+    while( mbedtls_mpi_bitlen( &ctx->D ) < (nbits / 2) );
```

Comment:
I think it should be `<= ( nbits / 2)` here: Since `bitlen` is 1-based and `D` is odd, we have `D > 2^( bitlen(D) - 1 )`, so we need `bitlen(D) - 1 >= nbits / 2`, i.e. `bitlen(D) > nbits / 2`, to infer `D > 2^( nbits / 2 )`.

Reply:
I'm not sure why it matters that `D` is odd.

I think the current comparison is correct:
```
1) 2^(nbits) > N                           // e.g. ∀ i ϵ [100b = 4d, 111b = 7d]: 2^3 > i
2) D >= 2^(dbits-1)                        // e.g. ∀ i ϵ [100b = 4d, 111b = 7d]: i >= 2^2

   want to show: D > N^0.5                 

3) 2^(nbits/2) > N^0.5                     // by (1)

4) D >= 2^(dbits-1) >? 2^(nbits/2) > N^0.5 // by (2), (3)

5) 2^(dbits-1) > 2^(nbits/2) =>  D > N^0.5 // by (4)
6) dbits - 1 > nbits/2       =>  D > N^0.5 
7) ! dbits - 1 <= nbits/2    =>  D > N^0.5 
8) ! dbits < nbits/2         =>  D > N^0.5 
```

Reply:
@jethrogb The standard requires that `D > N^0.5`, and the fact that `D` is odd allows to strengthen `(2) D >= 2^(dbits - 1)` to `D > 2^(dbits - 1)` - that's what I meant.

Thanks for spotting the detail about the rounding, I agree!